### ISSUE TITLE
Verify Cryptographic Signature in Elixir Server

### ISSUE THREAD

**Issue by lastmeta** [lastmeta/Satori#31]
TDLR:

The client generates a `public_key` and a `signature` of a `message`. Here's a valid example of those:

```
message = "message"
signature = "IFdGdzqI9fJJDbTASGbamHFQ7hlIT07jLALhYRhMrRY3QvKJuL2q8ojtwKsubj0EVtWJtlM8MbgmpvIaKwpjc04="
public_key = "034e328758d422ce5c3a17c15528df1216ef0eaf845147876eea82d139755129a0"
```

I need a function on the server that takes these two and spits out true or false. If the `signature` is verified using the `public_key` it should say `true`, if it is not verified it should say `false`. I already tried this using erlang's `crypto` package but it gives the wrong answer (says `false` when it should say `true`):

```
:crypto.verify(
  :ecdsa,
  :sha256,
  "message",
  Base.decode64("IFdGdzqI9fJJDbTASGbamHFQ7hlIT07jLALhYRhMrRY3QvKJuL2q8ojtwKsubj0EVtWJtlM8MbgmpvIaKwpjc04="),
  [Base.decode16("034e328758d422ce5c3a17c15528df1216ef0eaf845147876eea82d139755129a0", case: :lower), :secp256k1])
false
```
So I must be using it incorrectly or the client creates a special kind of signature. I don't know. I need someone to figure out how to "cryptographically verify" a signed message for the way the client produces a signature.

FULL DETAILS:

We need the ability to verify a blockchain-styled key-pair cryptographic signature on an Elixir server so that it can authenticate a Python Client through message signing.

We already have a wallet that can generate keys and sign messages at https://github.com/lastmeta/Satori/tree/main/node/satori/lib/wallet

For example:
```
>>> import satori.wallet as wallet
>>> w = wallet.Wallet()
>>> w.generate()
>>> w
Wallet(
    publicKey: 034e328758d422ce5c3a17c15528df1216ef0eaf845147876eea82d139755129a0
    privateKey: L16mQdokXV8hVxGc8su2z9xNkavWPdGGM92B6nt2bMj8yhjTCF8Z
    ...)
>>> sig = w.sign("message")
b'IFdGdzqI9fJJDbTASGbamHFQ7hlIT07jLALhYRhMrRY3QvKJuL2q8ojtwKsubj0EVtWJtlM8MbgmpvIaKwpjc04='
>>> w.verify("message", sig) 
True
```

We are trying to write the functionality on the server that verifies a signed message (like the one in the example). Upon connection the client will provide it's `publicKey` and the server will provide a `"message"` which the client then signs and returns the `signature`, which the server can then verify was signed correctly, authenticating the client.

Unfortunately, verifying the signature has been more difficult than expected, so that is the functionality we need help with on the Elixir server.

These two links describe our failed attempts and what we learned by them:
https://elixirforum.com/t/verifying-blockchain-crypto-signature-elixir-erlang/49226/5
https://stackoverflow.com/questions/73171510/verify-blockchain-message-signature-in-elixir

Notice, this attempt fails:
```
:crypto.verify(
  :ecdsa,
  :sha256,
  "message",
  Base.decode64!("IFdGdzqI9fJJDbTASGbamHFQ7hlIT07jLALhYRhMrRY3QvKJuL2q8ojtwKsubj0EVtWJtlM8MbgmpvIaKwpjc04="),
  [Base.decode16!("034e328758d422ce5c3a17c15528df1216ef0eaf845147876eea82d139755129a0"), :secp256k1])
false
```

In short, it seems the key pair generated by erlang's `:crypto` package doesn't match (in length) the key pairs generated by python's `python-ravencoinlib` (which is a fork of `python-bitcoinlib`) package. Perhaps this is because the encodings are different, we don't know where the problem lies.

**We need a function in Elixir that can verify the signature produced by the client:**

```
# myCrypto.verify(message, sig, pubkey) -> {:ok, true} | {:ok, false} | {:err, msg}

# example: 
#myCrypto.verify(
#  "message", 
#  "IFdGdzqI9fJJDbTASGbamHFQ7hlIT07jLALhYRhMrRY3QvKJuL2q8ojtwKsubj0EVtWJtlM8MbgmpvIaKwpjc04=",
#  "034e328758d422ce5c3a17c15528df1216ef0eaf845147876eea82d139755129a0")
#{:ok, true}
```

**Comment by abcpro1** [comment]
I think I figured out how the ravencoin signing works. I will post some code after I take a look at Elixir's crypto library.
By the way there is a bug [here](https://github.com/lastmeta/Satori/blob/c141a7ea725d0e00c334fe55858d5e0e599fb849/node/satori/lib/wallet/sign.py#L18-L20) which needs to be fixed.

**Comment by lastmeta** [comment]
> By the way there is a bug [here](https://github.com/lastmeta/Satori/blob/c141a7ea725d0e00c334fe55858d5e0e599fb849/node/satori/lib/wallet/sign.py#L18-L20) which needs to be fixed.

Thanks! I actually found that bug today, and replace it, will push soon!

**Comment by abcpro1** [comment]
Hi @lastmeta. I have written the verification code elixir as specified in your description and tested it successfully. Should I just paste it here?

**Comment by lastmeta** [comment]
> Hi @lastmeta. I have written the verification code in elixir as specified in your description and tested it successfully. Should I just paste it here?

Yes I will test it as soon as you do.

**Comment by abcpro1** [comment]
Alright here it is
```elixir
defmodule Ravencoin do
  @magic "Raven Signed Message:\n"

  # serialize data
  @spec serialize(binary) :: binary
  def serialize(data) do
    serialize_binary(@magic) <> serialize_binary(data)
  end

  # serialize binary
  @spec serialize_binary(binary) :: binary
  defp serialize_binary(data) do
    serialize_integer(byte_size(data)) <> data
  end

  # serialize variable sized integer
  @spec serialize_integer(integer) :: binary
  defp serialize_integer(value) do
    cond do
      value < 0 -> raise "varint must be a non-negative integer"
      value < 0xfd -> <<value>>
      value <= 0xffff -> <<0xfd, value::16-little>>
      value <= 0xffffffff -> <<0xfe, value::32-little>>
      true -> <<0xff, value::64-little>>
    end
  end
end

defmodule MyCrypto do
  @doc """
  Verify a signature.

  ## Parameters
  - message: the message in plain text
  - signature: signature encoded in base 64
  - public_key: the public key encoded in base 16
  """
  @spec verify!(binary, String.t, String.t) :: true | false
  def verify!(message, signature, public_key) do
    dersig = der_from_signature(signature |> Base.decode64!(ignore: :whitespace))
    public_key = [public_key |> Base.decode16!(case: :mixed), :secp256k1]
    doublehash = :crypto.hash(:sha256, :crypto.hash(:sha256, Ravencoin.serialize(message)))
    :crypto.verify(:ecdsa, :sha256, {:digest, doublehash}, dersig, public_key)
  end

  @doc "Ditto"
  @spec verify(binary, String.t, String.t) :: {:ok, true} | {:ok, false} | {:err, String.t}
  def verify(message, signature, public_key) do
    {:ok, verify!(message, signature, public_key)}
  rescue
    e -> {:error, Exception.message(e)}
  end

  # Convert singnature to DER.
  @spec der_from_signature(binary) :: binary
  defp der_from_signature(sig) do
    unless byte_size(sig) == 65 do
      raise "expected signature to be 65 bytes long, got #{byte_size(sig)} bytes"
    end
    <<_, r::binary-size(32), s::binary-size(32)>> = sig
    r = redo_der_integer_padding(r)
    s = redo_der_integer_padding(s)
    r_size = encode_asn1_length(byte_size(r))
    s_size = encode_asn1_length(byte_size(s))
    seq_size = 1 + byte_size(r_size) + byte_size(r) + 1 + byte_size(s_size) + byte_size(s)
    <<0x30>> <> encode_asn1_length(seq_size) <> <<0x2>> <> r_size <> r <> <<0x2>> <> s_size <> s
  end

  # Redo the padding of a DER integer.
  @spec redo_der_integer_padding(binary) :: binary
  defp redo_der_integer_padding(bn) do
    bn = String.trim_leading(bn, <<0>>)
    <<tip, _::binary>> = bn
    padding = if tip > 127, do: <<0>>, else: <<>>
    padding <> bn
  end

  # Encode ASN1 length.
  @spec encode_asn1_length(integer) :: binary
  defp encode_asn1_length(length) do
    if length < 128 do
      <<length>>
    else
      bytes = :binary.encode_unsigned(length, :little)
      <<128 + byte_size(bytes)>> <> bytes
    end
  end
end

unless length(System.argv()) == 3 do
  IO.puts "Verify a signature."
  IO.puts "usage: <message> <signature-in-base64> <private-key-in-base16>"
  System.halt(1)
end

[message, signature, public_key] = System.argv()

case MyCrypto.verify(message, signature, public_key) do
  {:ok, true} -> IO.puts "Signature Verified Successfully"
  {:ok, false} -> IO.puts "Signature Verification Failure"
  {:error, error} -> IO.puts "error: #{error}"
end
```
You can run it as a command line script whith
```sh
elixir script.exs $message $signature $publickey
```

**Comment by lastmeta** [comment]
@abcpro1 you're a genius. So I guess you mark it on bounty source as solved and I approve it, idk never used bountysource before but the bounty is yours!

**Comment by abcpro1** [comment]
Thank you.

**Comment by lastmeta** [comment]
@abcpro1 I'm impressed with your skills, and I'm sure we could use your help on this project as it evolves, if you'd like me to contact you direct for work just shoot me an email with your contact information at satorinetio@gmail.com, thanks!
